### ISSUE TITLE
not found: libdevice.compute_35.YY.bc

### ISSUE THREAD

**Issue by cpdurham** [AccelerateHS/accelerate-llvm#28]
cuda-9.1, lts-10.2, TITAN X (Pascal)

fresh repo on (stack new) with only accelerate, accelerate-llvm, accelerate-llvm-ptx added

getting:
```
    [30 of 42] Compiling Data.Array.Accelerate.LLVM.PTX.Compile.Libdevice.Load ( Data/Array/Accelerate/LLVM/PTX/Compile/Libdevice/Load.hs, .stack-work/dist/x86_64-linux/Cabal-2.0.1.0/build/Data/Array/Accelerate/LLVM/PTX/Compile/Libdevice/Load.o )
    
    /tmp/stack28093/accelerate-llvm-ptx-1.1.0.0/Data/Array/Accelerate/LLVM/PTX/Compile/Libdevice/Load.hs:106:23: error:
        • Exception when trying to run compile-time code:
            
    *** Internal error in package accelerate ***
    *** Please submit a bug report at https://github.com/AccelerateHS/accelerate/issues
    /tmp/stack28093/accelerate-llvm-ptx-1.1.0.0/Data/Array/Accelerate/LLVM/PTX/Compile/Libdevice/TH.hs:109:17: (libdevice): not found: libdevice.compute_35.YY.bc
    CallStack (from HasCallStack):
      error, called at ./Data/Array/Accelerate/LLVM/PTX/Compile/Libdevice/TH.hs:109:17 in accelerate-llvm-ptx-1.1.0.0-FTgu2aQ8ThhJWgN62BXqUo:Data.Array.Accelerate.LLVM.PTX.Compile.Libdevice.TH
          Code: libdeviceBitcode (Compute 3 5)
        • In the Template Haskell splice $$(libdeviceBitcode (Compute 3 5))
          In the expression: $$(libdeviceBitcode (Compute 3 5))
          In an equation for ‘libdevice_35_bc’:
              libdevice_35_bc = $$(libdeviceBitcode (Compute 3 5))
        |
    106 | libdevice_35_bc = $$( libdeviceBitcode (Compute 3 5) )
        |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

**Comment by tmcdonell** [comment]
Hey Charlie. Was this with `accelerate-llvm-ptx-1.1.0.0` by chance? I think this should be fixed in HEAD, but maybe they made additional changes between 9.0 and 9.1 I need to take care of? I was meant to do a patch release for this but totally forgot \\: (https://github.com/AccelerateHS/accelerate/issues/400)

**Comment by tmcdonell** [comment]
Should be fixed in [accelerate-llvm-ptx-1.1.0.1](https://hackage.haskell.org/package/accelerate-llvm-ptx-1.1.0.1). This should appear in the next lts-10.x series.

**Comment by ababkin** [comment]
Confirming that `accelerate-llvm-ptx-1.1.0.1` has suppressed the original issue, however it gives me the following when I'm running an executable (on `g2.2xlarge` AWS instance).
It works fine on native LLVM.

```
*** Internal error in package accelerate ***
*** Please submit a bug report at https://github.com/AccelerateHS/accelerate/issues
./Data/Array/Accelerate/Array/Remote/LRU.hs:239:27: (malloc): Remote memory exhausted
CallStack (from HasCallStack):
  error, called at ./Data/Array/Accelerate/Array/Remote/LRU.hs:239:27 in accelerate-1.1.1.0-1t0vFsrTYBAJJj9odZSnnG:Data.Array.Accelerate.Array.Remote.LRU
```

What other details should I supply?

**Comment by tmcdonell** [comment]
@ababkin From the error it seems like the GPU you are using just doesn't have enough on-board memory for whatever you are trying to do.

If you feel like this is a bug though (e.g., maybe accelerate is leaking memory, or you have confirmed that there is still memory available when the error fires), please do open another issue and we can discuss it (or, just to discuss this in more detail, and maybe we can find a solution for you...)

**Comment by ababkin** [comment]
Thanks @tmcdonell , I've played around with input size and indeed it looks plausible that it simply runs out of GPU memory.
I would like to know more about tooling available to profile the native/ptx memory usage so i can make informed decisions in my algorithms and see the feedback of my experimentations. Is there a handy guide for that or can you point me in the right direction?

**Comment by tmcdonell** [comment]
Kind of, but currently not in a way which relates back to the source program.

Probably the most useful debug mode options for you would be `-ddump-gc-stats`, which will give you a summary of the memory use at the end of your program. Running the ekg monitor to see the memory usage / allocate rate in real time might also help (this could be difficult for you if you are running remotely on AWS, as the default is to open up a port you can connect to with your web browser, but I believe ekg has various backends you can plug in to to get its data out). The `-ddump-gc` option is extremely verbose, but it'll give you the size and timestamp of every allocation (plus lots of other stuff; dump the output to a log and filter).

I'm interested in adding more tooling to help with this kind of stuff, but currently have very limited time. Suggestions for what in particular would be most useful for this sort of thing are very welcome.

**Comment by tmcdonell** [comment]
Also note that the way you structure the accelerate program will have an effect on memory usage; see for example [my comment here](https://github.com/AccelerateHS/accelerate/issues/411#issuecomment-356481589).

**Comment by ababkin** [comment]
Thanks again!
It's not obvious to me how to use the `-ddump-gc-stats` option. Tried giving it `+ACC -ddump-gc-stats -ACC` after the executable, but it tries to read it as CLI arguments. What should I put into my `stack.yaml` or the cabal file for this?

**Comment by tmcdonell** [comment]
Yeah, that is the way to use it. I have fixed that problem in HEAD, but if you are on the released versions still you should also be able to get around that by passing the flags through the environment variable `ACCELERATE_FLAGS` (iirc)

**Comment by tmcdonell** [comment]
uh, it occurs to me that `-ddump-gc-stats` probably isn't implemented in 1.1.0.1, only HEAD, but I guess you discovered that already. sorry for the misinformation!

**Comment by ababkin** [comment]
No problems Trevor, thanks a lot for your help
